themes/default/content/blog/managing-github-with-pulumi/index.md

```diff
@@ -0,0 +1,599 @@
+
+We gave each team their own config file with permission levels listing repository names:
+
+```yaml
+team: "animals"
+permissions:
+ - role: "admin"
+   repos:
+     - "capybaras"
+     - "owls"
+ - role: "maintain"
+   repos:
+ - role: "push"
+   repos:
+ - role: "read"
+   repos:
+```
+
+This can be read into our Pulumi program as follows:
+
+```go
+// TeamPermissions describes a github team and the levels of permissions available (i.e. admin, maintain)
+type TeamPermissions struct {
+  TeamName    string       `yaml:"team-name"`
+  Permissions []Permission `yaml:"permissions"`
+}
+
+// Permission is a single permission and a list of repos with that permission.
+type Permission struct {
+  Role  string   `yaml:"role"`
+  Repos []string `yaml:"repos"`
+}
+
+#...
+
+func setupTeamRepos(ctx *pulumi.Context, team *github.Team, teamName string) error {
+
+  repoFilePath, err := filepath.Abs("./team-repos/" + teamName + ".yaml")
+  if err != nil {
+     fmt.Println(err.Error())
+     return err
+  }
+  yamlFile, err := ioutil.ReadFile(repoFilePath)
+  if err != nil {
+     return err
+  }
+
+  var teamPerms TeamPermissions
+  err = yaml.Unmarshal(yamlFile, &teamPerms)
+  if err != nil {
+     return err
+  }
+  for _, permission := range teamPerms.Permissions {
+     for _, repo := range permission.Repos {
+        _, err := github.NewTeamRepository(ctx, fmt.Sprintf("%s-%s-%s", repo, teamName, permission.Role), &github.TeamRepositoryArgs{
+           Permission: pulumi.String(permission.Role),
+           Repository: pulumi.String(repo),
+           TeamId:     team.ID(),
+        })
+
+        if err != nil {
+           return err
+        }
+     }
+  }
+  return nil
+}
+```
+
+And the output of running `pulumi up`:
+
+```bash
+Updating (prod)
+
+     Type                            Name                      Status
+     pulumi:pulumi:Stack             team-mgmt-prod
+ +   ├─ github:index:TeamRepository  owls-animals-admin        created
+ +   ├─ github:index:TeamRepository  capybaras-animals-admin   created
+
+
+Resources:
+    + 2 created
+    2 changes. 4 unchanged
+```
+
+With this addition, we can now combine teams with zero access disruptions as follows:
+
+1. Create the new team in the org config
+2. Create a TeamRepository config for the new team with desired permission levels
+3. Open a pull request and let CI and Pulumi do the rest.
+
+Now we can develop standards around repository permissions based on teams’ roles and areas of responsibility, and we have tooling in place that can maintain these standards for everyone in the org to see. In fact, we can standardize the meaning of “code ownership” via access levels in this way - one centralized management location rather than asking your grandboss to dig through the UI for you.
```

Comment:
```suggestion
Now we can develop standards around repository permissions based on teams’ roles and areas of responsibility, and we have tooling in place that can maintain these standards for everyone in the org to see. In fact, we can standardize the meaning of “code ownership” via access levels in this way---one centralized management location rather than asking your grandboss to dig through the UI for you.
```